src/Cardano/Wallet/Binary.hs

```diff
@@ -0,0 +1,370 @@
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE DataKinds #-}
+{-# OPTIONS_GHC -fno-warn-name-shadowing #-}
+
+-- | These are (partial) CBOR decoders for blocks and block headers.
+-- Note that we ignore most of the block's and header's content and only
+-- retrieve the pieces of information relevant to us, wallet (we do assume a
+-- trusted node and therefore, we needn't to care about verifying signatures and
+-- blocks themselves).
+
+module Cardano.Wallet.Binary
+    ( decodeBlock
+    , decodeBlockHeader
+    ) where
+
+import           Prelude
+import           Control.Monad                    (void)
+import qualified Codec.CBOR.Decoding              as CBOR
+import qualified Codec.CBOR.Encoding              as CBOR
+import qualified Codec.CBOR.Read                  as CBOR
+import qualified Codec.CBOR.Write                 as CBOR
+import qualified Data.ByteString.Lazy             as BL
+import           Data.Word                        (Word16, Word64)
+import           Data.Set                         (Set)
+import qualified Data.Set                         as Set
+
+
+import Cardano.Wallet.Primitive
+import Cardano.Wallet.Binary.Helpers
+
+-- fixme: Temporary placeholder
+data TxWitness = TxWitness deriving (Show)
+
+decodeAddress :: CBOR.Decoder s Address
+decodeAddress = do
+    _ <- CBOR.decodeListLenCanonicalOf 2 -- CRC Protection Wrapper
+    tag <- CBOR.decodeTag -- Myterious hard-coded tag cardano-sl seems to so much like
+    bytes <- CBOR.decodeBytes -- Addr Root + Attributes + Type
+    crc <- CBOR.decodeWord32 -- CRC
+    -- NOTE 1:
+    -- Treating addresses as a blob here, so we just re-encode them as such
+    -- Ultimately for us, addresses are nothing more than a bunch of bytes that
+    -- we display in a Base58 format when we have too.
+    --
+    -- NOTE 2:
+    -- We may want to check the CRC at this level as-well... maybe not.
+    return $ Address $ CBOR.toStrictByteString $ mempty
+        <> CBOR.encodeListLen 2
+        <> CBOR.encodeTag tag
+        <> CBOR.encodeBytes bytes
+        <> CBOR.encodeWord32 crc
+
+
+decodeAttributes :: CBOR.Decoder s ((), CBOR.Encoding)
+decodeAttributes = do
+    _ <- CBOR.decodeMapLenCanonical -- Empty map of attributes
+    return ((), CBOR.encodeMapLen 0)
+
+decodeBlock :: CBOR.Decoder s Block
+decodeBlock = do
+    CBOR.decodeListLenCanonicalOf 2
+    t <- CBOR.decodeWordCanonical
+    case t of
+        0 -> do -- Genesis Block
+            _ <- CBOR.decodeListLenCanonicalOf 3
+            header <- decodeGenesisBlockHeader
+            -- NOTE
+            -- We don't decode the body of genesis block because we don't
+            -- need it (see @decodeGenesisBlockBody@). Genesis blocks occurs at
+            -- every epoch boundaries and contains various information about
+            -- protocol updates, slot leaders elections and delegation.
```

Comment:
"We don't decode the body of genesis block" -> "We don't decode the body of a genesis block"
"Genesis blocks occurs at" -> "Genesis blocks occur at"
"every epoch boundaries" -> "every epoch boundary"
"and contains various information" -> "and contain various pieces of information"
"slot leaders elections" -> "slot leader elections"

Reply:
Fixed